taraxa/state/dpos/precompiled/dpos_contract.go

```diff
@@ -0,0 +1,1022 @@
+		// TODO: question: how can we reset validator's rewards pool after singl delegator claim rewards ?
+		validator.RewardsPool = bigutil.Big0
+		validator.LastUpdated = block
+		state.Count++
+		self.validators.ModifyValidator(&args.Validator, validator)
+	}
+
+	old_state := self.state_get_and_decrement(args.Validator[:], BlockToBytes(delegation.LastUpdated))
+	reward := bigutil.Sub(state.RwardsPer1Stake, old_state.RwardsPer1Stake)
+	// TODO: question: how is it possible that in case state == nil, we give delegator some rewards but we dont adjust validator's rewards pool ???
+	ctx.CallerAccount.AddBalance(bigutil.Mul(reward, delegation.Stake))
+	delegation.LastUpdated = block
+	self.delegations.ModifyDelegation(ctx.CallerAccount.Address(), &args.Validator, delegation)
+
+	state.Count++
+	self.state_put(&state_k, state)
+
+	return nil
+}
+
+func (self *Contract) claimCommissionRewards(ctx vm.CallFrame, block types.BlockNum) error {
+	validator_address := ctx.CallerAccount.Address()
+	validator := self.validators.GetValidator(validator_address)
+	if validator == nil {
+		return ErrNonExistentValidator
+	}
+
+	state, state_k := self.state_get(ctx.CallerAccount.Address()[:], BlockToBytes(block))
+	if state == nil {
+		old_state := self.state_get_and_decrement(ctx.CallerAccount.Address()[:], BlockToBytes(validator.LastUpdated))
+		state = new(State)
+		state.RwardsPer1Stake = bigutil.Add(old_state.RwardsPer1Stake, bigutil.Div(validator.RewardsPool, validator.TotalStake))
+		validator.RewardsPool = bigutil.Big0
+		validator.LastUpdated = block
+		state.Count++
+	}
+
+	ctx.CallerAccount.AddBalance(validator.CommissionRewardsPool)
+	validator.CommissionRewardsPool = bigutil.Big0
+
+	self.validators.ModifyValidator(validator_address, validator)
+	self.state_put(&state_k, state)
+
+	return nil
+}
+
+func (self *Contract) registerValidator(ctx vm.CallFrame, block types.BlockNum, args RegisterValidatorArgs) error {
+	validator_address := ctx.CallerAccount.Address()
+
+	if self.validators.ValidatorExists(validator_address) {
+		return ErrExistentValidator
+	}
+
+	delegation := self.delegations.GetDelegation(validator_address, validator_address)
+	if delegation != nil {
+		// This could happen only due some serious logic bug
+		panic("registerValidator: delegation already exists")
+	}
+
+	state, state_k := self.state_get(validator_address[:], BlockToBytes(block))
+	if state != nil {
+		return ErrBrokenState
+	}
+
+	// TODO: limit size of description & endpoint - should be very small
+
+	ctx.Account.SubBalance(ctx.Value) // TODO how to get correct value?
+
+	state = new(State)
+	state.RwardsPer1Stake = bigutil.Big0
+
+	// Creates validator related objects in storage
+	self.validators.CreateValidator(validator_address, block, ctx.Value, args.Commission, args.Description, args.Endpoint)
+	state.Count++
+
+	if ctx.Value.Cmp(bigutil.Big0) == 1 {
```

Comment:
this is not good. It is required for each validator self delegates some min. amount to himself during registration(he can later withdraw). We should put this constant into the config (& validate the config value so it cannot be == 0) and then check it here

Reply:
so let's create task for that. Currently we allowed this

Reply:
why dont we just fix it ? It is relatively simple change

Reply:
because this is correct use-case for old DPOS, that validator stakes 0 and he has only delegations. This should be part of the config if it is allowed or not